[FLINK-24897] Enable Application/Per-job mode on YARN to use usrlib

bgeng777 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request aims at enabling application/per-job mode on YARN to use usrlib easier. 


## Brief change log
  - Make `usrlib`(if exists) under FLINK_HOME be uploaded to the cluster automatically for both Application and Per-job mode on YARN.  If we add `usrlib`  in ship files again or specify `usrlib` in `yarn.provided.lib.dirs`, exception will be thrown. 
  - Jars in `usrlib` will be loaded by user classloader only when `yarn.per-job-cluster.include-user-jar` is DISABLED. 



## Verifying this change


This change added tests and can be verified as follows:
  - Add tests that validate the shipping logic
  - Manually verified the classloader is chosen correctly when `yarn.per-job-cluster.include-user-jar` is DISABLED

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no
  - 
## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs(but need improvement)

bgeng777 commented:
@flinkbot run azure

bgeng777 commented:
@flinkbot run azure

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:
```diff
@@ -438,6 +446,18 @@ private static boolean isPlugin(Path path) {
         return false;
     }
 
+    private static boolean isUsrLib(Path path) {
```
I suggest to introduce the `isUsrLibDirIncludedInProvidedLib` instead and then `checkArgument` in the constructor of `YarnApplicationFileUploader`. Because the files in the provided lib will be relativized, so we just need to check the name of provided lib.

```
    private boolean isUsrLibDirIncludedInProvidedLib(final List<Path> providedLibDirs)
            throws IOException {
        for (Path path : providedLibDirs) {
            final FileStatus fileStatus = fileSystem.getFileStatus(path);
            if (fileStatus.isDirectory()
                    && ConfigConstants.DEFAULT_FLINK_USR_LIB_DIR.equals(path.getName())) {
                return true;
            }
        }
        return false;
    }
```

bgeng777 replied:
Make sense. Fixed in ef3f0104fc31def407b74b4698d5f7dd0404a333

wangyang0918 reviewed (changes requested):
@bgeng777 Thanks for creating this PR. I think your proposal makes sense and could work. And I have left some comments about the implementation. Please have a look.

bgeng777 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -1685,6 +1699,35 @@ void addLibFoldersToShipFiles(Collection<File> effectiveShipFiles) {
+        if (usrLibDir.isPresent()) {
+            File usrLibDirFile = usrLibDir.get();
+            if (usrLibDirFile.isDirectory()) {
+                checkArgument(
```
Hi  Yang, thanks for the feedback. In fact, my first version of codes is pretty like your code snippet. 
But after thinking more carefully, I use my current implementation (including writing my own getLocalUsrLibDirectory()) for the following reason:
1. AFAIK, YarnClusterDescriptor is executed in the flink client side while ClusterEntryPointUtil is executed on the remote cluster side. If I use the `ClusterEntrypointUtils.tryFindUserLibDirectory()` in client side, due to its impl, the user must specify the FLINK_LIB_DIR correctly. But IIUC, we do not have such requirement before when using YARN. 
2. In my `getLocalUsrLibDirectory()`, I use the FLINK_HOME as a pivot to find the `usrlib`. This is not so perfect but comparing with FLINK_LIB_DIR, FLINK_HOME may be more widely used.

In summary, I bet the key question is if we should ask the action of finding `usrlib` to adopt same logic(using FLINK_LIB_DIR or use current working dir) in both client and cluster side.  Do you have any suggestion?

wangyang0918 replied:
Except for the naming, I do not think the `ClusterEntrypointUtils.tryFindUserLibDirectory()` should only work in the cluster side. If a user configure a customized `FLINK_LIB_DIR`, then he/she should also have the `usrlib` in such location. I think it is reasonable.

For the naming, maybe we could have the `UserLibUtils.java`, which contains all the utility methods for `usrlib`.

bgeng777 replied:
Agree with the naming and the `FLINK_LIB_DIR` clarification.
Just one follow-up: I know that `FLINK_LIB_DIR` will be set by `config.sh` if it is empty when we run `flink run ....`.
But when YARN run commands like `/bin/bash -c /usr/lib/jvm/java-1.8.0/bin/java -Xmx1073741824 -Xms1073741824 -XX:MaxMetaspaceSize=268435456 -Dlog.file=/mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.log -Dlog4j.configuration=file:log4j.properties -Dlog4j.configurationFile=file:log4j.properties org.apache.flink.yarn.entrypoint.YarnJobClusterEntrypoint -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=201326592b -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=1073741824b -D jobmanager.memory.jvm-overhead.max=201326592b 1> /mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.out 2> /mnt/disk1/log/hadoop-yarn/containers/application_1643263086460_0041/container_1643263086460_0041_01_000001/jobmanager.err`, I think the `bin/config.sh` will not executed? If I am wrong, please correct me. Thanks!

wangyang0918 replied:
What I mean is about `flink run/run-applicaton` on the client side. For the JobManager and TaskManager start commands, you are right that `config.sh` will not be executed. So I think we could have a dedicated ticket to fix the issue you comment under [FLINK-24897](https://issues.apache.org/jira/browse/FLINK-24897?focusedCommentId=17483689&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17483689).

bgeng777 replied:
No problem. In this pr, `ClusterEntrypointUtils.tryFindUserLibDirectory()` is used since 986fff4543fbfd5016f087c53533b395af8c166f.

bgeng777 commented on flink-yarn/pom.xml:
```diff
@@ -130,6 +130,11 @@ under the License.
 				</exclusion>
 			</exclusions>
 		</dependency>
+		<dependency>
```
My bad. Removed.

bgeng777 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -910,14 +910,28 @@ private ApplicationReport startAppMaster(
 
         // Upload and register user jars
-        final List<String> userClassPaths =
+        List<String> userClassPaths =
```
Because we need to check if usrlib exists later. If usrlib exists, we need `userClassPaths.addAll(usrLibClassPaths);` so we can reuse current code to make other options o YarnConfigOptions.UserJarInclusion work as expected.

wangyang0918 replied:
Even though we could still have the `final`.

bgeng777 replied:
Fix in 986fff4543fbfd5016f087c53533b395af8c166f

bgeng777 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -1799,4 +1842,27 @@ public static void logDetachedClusterInformation(
                 yarnApplicationId);
     }
+
+    /**
```
See https://github.com/apache/flink/pull/18531/files/5eb98df7422ffca67eb4d258189ff1eadb36731e#discussion_r801320099

bgeng777 replied:
Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

bgeng777 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:
```diff
@@ -335,7 +335,7 @@ public YarnLocalResourceDescriptor uploadFlinkDist(final Path localJarPath)
      * @return list of class paths with the file name
      */
-    List<String> registerProvidedLocalResources() {
+    List<String> registerProvidedLocalResources(boolean shouldUsrLibExistInRemote) {
```
Yes, I totally agree with the assumption and that it just have the same semantic with ship files. 
But in ship files, we will check if `usrlib` exists in ship files and if so and `INCLUDE_USER_JAR` is DISABLED, we will throw exception due to our previous discussion. 
The reason for me to adding such code is for this case: users have local `usrlib` and in the remote dirs specified by `provided.lib.dirs`, there is also a remote `usrlib`. I want to get rid of  such case as it could lead to overriding files in YARN cache's `usrlib`.

wangyang0918 replied:
I think no matter the `INCLUDE_USER_JAR` is configured, we should always not include `usrlib` in the user shipped files. This is also true for the provided lib.

bgeng777 replied:
It makes sense to me. I will update the JIRA as well to emphasize  we should always not include `usrlib` in the user shipped files.   Codes are much easier with this.
Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

bgeng777 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -872,14 +872,14 @@ private ApplicationReport startAppMaster(
             jobGraph.writeUserArtifactEntriesToConfiguration();
         }
-
+        boolean isLocalUsrLibExists = getLocalUsrLibDirectory().isPresent();
```
Fixed in ef3f0104fc31def407b74b4698d5f7dd0404a333.

bgeng777 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +700,41 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+     * shipped.
+     */
+    @Test
+    public void testShipUsrLibWithExistingLocalUsrLib() throws IOException {
```
Fixed in 986fff4543fbfd5016f087c53533b395af8c166f

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +699,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+    @Test
+    public void testShipUsrLib() throws IOException {
+        final Map<String, String> oldEnv = System.getenv();
+        Map<String, String> env = new HashMap<>(1);
```
Please use `final` if possible.

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +699,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+        try (YarnClusterDescriptor descriptor = createYarnClusterDescriptor()) {
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            Assert.assertFalse(effectiveShipFiles.contains(usrLibFile));
```
Use `assertThat` is always recommended.

```
assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
```

wangyang0918 reviewed (changes requested):
Thanks for addressing the comments. It looks almost good enough and I just left some minor suggestions.

One more thing about the config option `yarn.per-job-cluster.include-user-jar`. Could we 
 rename it to `yarn.classpath.include-user-jar` and deprecate the old one?

bgeng777 commented:
Hi @wangyang0918 thanks again for the detailed review. Above comments has been resolved in latest commit .
For the renaming of config option `yarn.per-job-cluster.include-user-jar`, I have adopted your advice, added `withDeprecatedKeys` and updated comments in the code.
 But I want to check if I should update markdown docs in this PR as well. It is not much work though I checked some Flink PRs and it seems docs tend to be modified in a doc revision PR. Any guidance on which one is recommended is appreciated.

wangyang0918 commented:
We already have a PR #18450 to document how to use `usrlib`. But you still need to regenerate the config docs.
```
mvn package -Dgenerate-config-docs -pl flink-docs -am -nsu -DskipTests
```

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -876,7 +874,6 @@ private ApplicationReport startAppMaster(
         if (providedLibDirs == null || providedLibDirs.isEmpty()) {
             addLibFoldersToShipFiles(systemShipFiles);
         }
-
```
Unnecessary change.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -918,6 +915,18 @@ private ApplicationReport startAppMaster(
                         LocalResourceType.FILE);
 
+        // usrlib will be automatically shipped if it exists.
+        final Set<File> usrLibShipFiles = new HashSet<>();
```
Hmm. Maybe I did not make myself clear. I do not mean to remove the condition check.

```
if (ClusterEntrypointUtils.tryFindUserLibDirectory().isPresent()) {
    ... ...
}
```

bgeng777 replied:
I remove this check because I double checked the codes and it should work well as we will do this check in addUsrLibFolderToShipFiles. 
For readability and less object allocation/function call , I agree to get it back.

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnApplicationFileUploaderTest.java:
```diff
@@ -100,6 +105,35 @@ public void testRegisterProvidedLocalResourcesWithDuplication() throws IOExcepti
+                                        providedLibDirs,
+                                        ApplicationId.newInstance(0, 0),
+                                        DFSConfigKeys.DFS_REPLICATION_DEFAULT))
+                .isInstanceOf(RuntimeException.class)
```
Maybe we could use a more specific `IllegalArgumentException` here.

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
+            assertThat(effectiveShipFiles, not(containsInAnyOrder(usrLibFile)));
```
I do not think we still need this assert. The above `containsInAnyOrder` could cover such case.

bgeng777 replied:
The `assertThat(effectiveShipFiles, not(containsInAnyOrder(usrLibFile)));` just follows the check logic in testEnvironmentDirectoryShipping to make sure single is not added.

bgeng777 replied:
You are right. Our `tryFindUserLibDirectory` will only return a directory.

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+        try (YarnClusterDescriptor descriptor = createYarnClusterDescriptor()) {
+            Set<File> effectiveShipFiles = new HashSet<>();
+            descriptor.addUsrLibFolderToShipFiles(effectiveShipFiles);
+            assertThat(effectiveShipFiles, containsInAnyOrder(usrLibFolder));
```
Please use `org.hamcrest.MatcherAssert.assertThat` instead.

wangyang0918 commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterDescriptorTest.java:
```diff
@@ -699,6 +701,32 @@ public void testDisableSystemClassPathIncludeUserJarAndWithIllegalShipDirectoryN
+    public void testShipUsrLib() throws IOException {
+        final Map<String, String> oldEnv = System.getenv();
+        final Map<String, String> env = new HashMap<>(1);
+        File homeFolder = temporaryFolder.newFolder().getAbsoluteFile();
```
I believe `homeFolder` `libFolder` `usrLibFolder` `usrLibFile` `effectiveShipFiles` could also be final.

wangyang0918 reviewed (approved):
LTGM. +1

I will rebase master, squash the commits and then merge this PR once the azure pipeline gives a pass.

wangyang0918 commented:
@flinkbot run azure

bgeng777 commented:
The azure test pipeline is broken for now and no e2e tests can succeed for all tests for now. Will run azure test tomorrow if it works.

bgeng777 commented:
Rerun the Flink CI in my own Azure pipeline and get a pass:
https://dev.azure.com/samuelgeng7/Flink/_build/results?buildId=74&view=results

wangyang0918 commented:
Merging this PR now.
